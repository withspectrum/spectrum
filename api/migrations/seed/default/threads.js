// @flow
const { fromPlainText, toJSON } = require('../../../../shared/draft-utils');
const constants = require('./constants');
const {
  DATE,
  BRIAN_ID,
  MAX_ID,
  BRYN_ID,
  SPECTRUM_GENERAL_CHANNEL_ID,
  PRIVATE_GENERAL_CHANNEL_ID,
  SPECTRUM_PRIVATE_CHANNEL_ID,
  DELETED_COMMUNITY_DELETED_CHANNEL_ID,
  MODERATOR_CREATED_CHANNEL_ID,
  DELETED_COMMUNITY_ID,
  SPECTRUM_COMMUNITY_ID,
  PRIVATE_COMMUNITY_ID,
  SPECTRUM_ARCHIVED_CHANNEL_ID,
} = constants;

module.exports = [
  {
    id: 'thread-1',
    createdAt: new Date(DATE),
    creatorId: BRIAN_ID,
    channelId: SPECTRUM_GENERAL_CHANNEL_ID,
    communityId: SPECTRUM_COMMUNITY_ID,
    isPublished: true,
    isLocked: false,
    type: 'DRAFTJS',
    content: {
      title: 'The first thread! ðŸŽ‰',
      body: JSON.stringify(
        toJSON(fromPlainText('This is it, we got a thread here'))
      ),
    },
    attachments: [],
    edits: [
      {
        timestamp: new Date(DATE),
        content: {
          title: 'The first thread! ðŸŽ‰',
          body: JSON.stringify(
            toJSON(fromPlainText('This is it, we got a thread here'))
          ),
        },
      },
    ],
    modifiedAt: new Date(DATE),
    lastActive: new Date(DATE),
    messageCount: 4,
    reactionCount: 0,
  },
  {
    id: 'thread-2',
    createdAt: new Date(DATE + 1),
    creatorId: MAX_ID,
    channelId: SPECTRUM_GENERAL_CHANNEL_ID,
    communityId: SPECTRUM_COMMUNITY_ID,
    isPublished: true,
    isLocked: false,
    type: 'DRAFTJS',
    content: {
      title: 'Another thread',
      body: JSON.stringify(
        toJSON(fromPlainText('This is just another thread'))
      ),
    },
    attachments: [],
    edits: [
      {
        timestamp: new Date(DATE + 1),
        content: {
          title: 'Another thread',
          body: JSON.stringify(
            toJSON(fromPlainText('This is just another thread'))
          ),
        },
      },
    ],
    modifiedAt: new Date(DATE + 1),
    lastActive: new Date(DATE + 1),
    messageCount: 4,
    reactionCount: 0,
  },
  {
    id: 'thread-3',
    createdAt: new Date(DATE + 2),
    creatorId: BRYN_ID,
    channelId: SPECTRUM_GENERAL_CHANNEL_ID,
    communityId: SPECTRUM_COMMUNITY_ID,
    isPublished: true,
    isLocked: false,
    type: 'DRAFTJS',
    content: {
      title: 'Yet another thread',
      body: JSON.stringify(
        toJSON(fromPlainText('This is just another thread'))
      ),
    },
    attachments: [],
    edits: [
      {
        timestamp: new Date(DATE + 2),
        content: {
          title: 'Yet another thread',
          body: JSON.stringify(
            toJSON(fromPlainText('This is just another thread'))
          ),
        },
      },
    ],
    modifiedAt: new Date(DATE + 2),
    lastActive: new Date(DATE + 2),
    messageCount: 0,
    reactionCount: 0,
  },

  {
    id: 'thread-4',
    createdAt: new Date(DATE),
    creatorId: BRIAN_ID,
    channelId: SPECTRUM_PRIVATE_CHANNEL_ID,
    communityId: SPECTRUM_COMMUNITY_ID,
    isPublished: true,
    isLocked: false,
    type: 'DRAFTJS',
    content: {
      title: 'The first thread! ðŸŽ‰',
      body: JSON.stringify(
        toJSON(fromPlainText('This is it, we got a thread here'))
      ),
    },
    attachments: [],
    edits: [
      {
        timestamp: new Date(DATE),
        content: {
          title: 'The first thread! ðŸŽ‰',
          body: JSON.stringify(
            toJSON(fromPlainText('This is it, we got a thread here'))
          ),
        },
      },
    ],
    modifiedAt: new Date(DATE),
    lastActive: new Date(DATE),
    messageCount: 0,
    reactionCount: 0,
  },
  {
    id: 'thread-5',
    createdAt: new Date(DATE + 1),
    creatorId: MAX_ID,
    channelId: SPECTRUM_PRIVATE_CHANNEL_ID,
    communityId: SPECTRUM_COMMUNITY_ID,
    isPublished: true,
    isLocked: false,
    type: 'DRAFTJS',
    content: {
      title: 'Another thread',
      body: JSON.stringify(
        toJSON(fromPlainText('This is just another thread'))
      ),
    },
    attachments: [],
    edits: [
      {
        timestamp: new Date(DATE + 1),
        content: {
          title: 'Another thread',
          body: JSON.stringify(
            toJSON(fromPlainText('This is just another thread'))
          ),
        },
      },
    ],
    modifiedAt: new Date(DATE + 1),
    lastActive: new Date(DATE + 1),
    messageCount: 0,
    reactionCount: 0,
  },
  {
    id: 'thread-6',
    createdAt: new Date(DATE + 2),
    creatorId: BRYN_ID,
    channelId: SPECTRUM_PRIVATE_CHANNEL_ID,
    communityId: SPECTRUM_COMMUNITY_ID,
    isPublished: true,
    isLocked: false,
    type: 'DRAFTJS',
    content: {
      title: 'Yet another thread',
      body: JSON.stringify(
        toJSON(fromPlainText('This is just another thread'))
      ),
    },
    attachments: [],
    edits: [
      {
        timestamp: new Date(DATE + 2),
        content: {
          title: 'Yet another thread',
          body: JSON.stringify(
            toJSON(fromPlainText('This is just another thread'))
          ),
        },
      },
    ],
    modifiedAt: new Date(DATE + 2),
    lastActive: new Date(DATE + 2),
    messageCount: 0,
    reactionCount: 0,
  },
  {
    id: 'thread-7',
    createdAt: new Date(DATE + 2),
    creatorId: BRYN_ID,
    channelId: DELETED_COMMUNITY_DELETED_CHANNEL_ID,
    communityId: DELETED_COMMUNITY_ID,
    isPublished: true,
    isLocked: false,
    type: 'DRAFTJS',
    content: {
      title: 'Yet another thread',
      body: JSON.stringify(
        toJSON(fromPlainText('This is just another thread'))
      ),
    },
    attachments: [],
    edits: [
      {
        timestamp: new Date(DATE + 2),
        content: {
          title: 'Yet another thread',
          body: JSON.stringify(
            toJSON(fromPlainText('This is just another thread'))
          ),
        },
      },
    ],
    modifiedAt: new Date(DATE + 2),
    lastActive: new Date(DATE + 2),
    deletedAt: new Date(DATE + 3),
    messageCount: 0,
    reactionCount: 0,
  },
  {
    id: 'thread-8',
    createdAt: new Date(DATE + 2),
    creatorId: BRYN_ID,
    channelId: SPECTRUM_ARCHIVED_CHANNEL_ID,
    communityId: SPECTRUM_COMMUNITY_ID,
    isPublished: true,
    isLocked: false,
    type: 'DRAFTJS',
    content: {
      title: 'Yet another thread',
      body: JSON.stringify(
        toJSON(fromPlainText('This is just another thread'))
      ),
    },
    attachments: [],
    edits: [
      {
        timestamp: new Date(DATE + 2),
        content: {
          title: 'Yet another thread',
          body: JSON.stringify(
            toJSON(fromPlainText('This is just another thread'))
          ),
        },
      },
    ],
    modifiedAt: new Date(DATE + 2),
    lastActive: new Date(DATE + 2),
    messageCount: 0,
    reactionCount: 0,
  },
  {
    id: 'thread-9',
    createdAt: new Date(DATE + 2),
    creatorId: BRYN_ID,
    channelId: SPECTRUM_GENERAL_CHANNEL_ID,
    communityId: SPECTRUM_COMMUNITY_ID,
    isPublished: true,
    isLocked: true,
    type: 'DRAFTJS',
    content: {
      title: 'Yet another thread',
      body: JSON.stringify(
        toJSON(fromPlainText('This is just another thread'))
      ),
    },
    attachments: [],
    edits: [
      {
        timestamp: new Date(DATE + 2),
        content: {
          title: 'Yet another thread',
          body: JSON.stringify(
            toJSON(fromPlainText('This is just another thread'))
          ),
        },
      },
    ],
    modifiedAt: new Date(DATE + 2),
    lastActive: new Date(DATE + 2),
  },

  {
    id: 'thread-10',
    createdAt: new Date(DATE + 2),
    creatorId: BRYN_ID,
    channelId: MODERATOR_CREATED_CHANNEL_ID,
    communityId: SPECTRUM_COMMUNITY_ID,
    isPublished: true,
    isLocked: false,
    type: 'DRAFTJS',
    content: {
      title: 'Yet another thread',
      body: JSON.stringify(
        toJSON(fromPlainText('This is just another thread'))
      ),
    },
    attachments: [],
    edits: [
      {
        timestamp: new Date(DATE + 2),
        content: {
          title: 'Yet another thread',
          body: JSON.stringify(
            toJSON(fromPlainText('This is just another thread'))
          ),
        },
      },
    ],
    modifiedAt: new Date(DATE + 2),
    lastActive: new Date(DATE + 2),
    messageCount: 0,
    reactionCount: 0,
  },
  {
    id: 'thread-11',
    createdAt: new Date(DATE + 2),
    creatorId: BRYN_ID,
    channelId: SPECTRUM_GENERAL_CHANNEL_ID,
    communityId: SPECTRUM_COMMUNITY_ID,
    isPublished: true,
    isLocked: true,
    type: 'DRAFTJS',
    content: {
      title: 'Deleted thread',
      body: JSON.stringify(toJSON(fromPlainText('This is a deleted thread'))),
    },
    attachments: [],
    edits: [
      {
        timestamp: new Date(DATE + 2),
        content: {
          title: 'Deleted thread',
          body: JSON.stringify(
            toJSON(fromPlainText('This is a deleted thread'))
          ),
        },
      },
    ],
    modifiedAt: new Date(DATE + 2),
    lastActive: new Date(DATE + 2),
    deletedAt: new Date(DATE + 3),
    messageCount: 0,
    reactionCount: 0,
  },

  {
    id: 'thread-12',
    createdAt: new Date(DATE + 2),
    creatorId: BRYN_ID,
    channelId: DELETED_COMMUNITY_DELETED_CHANNEL_ID,
    communityId: DELETED_COMMUNITY_ID,
    isPublished: true,
    isLocked: false,
    type: 'DRAFTJS',
    content: {
      title: 'Yet another thread',
      body: JSON.stringify(
        toJSON(fromPlainText('This is just another thread'))
      ),
    },
    attachments: [],
    edits: [
      {
        timestamp: new Date(DATE + 2),
        content: {
          title: 'Yet another thread',
          body: JSON.stringify(
            toJSON(fromPlainText('This is just another thread'))
          ),
        },
      },
    ],
    modifiedAt: new Date(DATE + 2),
    lastActive: new Date(DATE + 2),
    deletedAt: new Date(DATE),
    messageCount: 0,
    reactionCount: 0,
  },

  {
    id: 'thread-13',
    createdAt: new Date(DATE + 2),
    creatorId: MAX_ID,
    channelId: PRIVATE_GENERAL_CHANNEL_ID,
    communityId: PRIVATE_COMMUNITY_ID,
    isPublished: true,
    isLocked: false,
    type: 'DRAFTJS',
    content: {
      title: 'Yet another thread',
      body: JSON.stringify(
        toJSON(fromPlainText('This is just another thread'))
      ),
    },
    attachments: [],
    edits: [
      {
        timestamp: new Date(DATE + 2),
        content: {
          title: 'Yet another thread',
          body: JSON.stringify(
            toJSON(fromPlainText('This is just another thread'))
          ),
        },
      },
    ],
    modifiedAt: new Date(DATE + 2),
    lastActive: new Date(DATE + 2),
    messageCount: 0,
    reactionCount: 0,
  },
];
