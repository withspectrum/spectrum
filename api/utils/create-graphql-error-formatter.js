// @flow
const debug = require('debug')('api:utils:error-formatter');
import Raven from 'shared/raven';
import { IsUserError } from './UserError';
import type { GraphQLError } from 'graphql';
import { Request } from 'api/index';

const queryRe = /\s*(query|mutation)[^{]*/;

const collectQueries = query => {
  if (!query) return 'No query';
  return query
    .split('\n')
    .map(line => {
      const m = line.match(queryRe);
      return m ? m[0].trim() : '';
    })
    .filter(line => !!line)
    .join('\n');
};

const errorPath = error => {
  if (!error.path) return '';
  return error.path
    .map((value, index) => {
      if (!index) return value;
      return typeof value === 'number' ? `[${value}]` : `.${value}`;
    })
    .join('');
};

const logGraphQLError = (req, error) => {
  debug('---GraphQL Error---');
  debug(error);
  if (req) {
    debug(collectQueries(req.body.query));
    debug('variables', JSON.stringify(req.body.variables || {}));
  }
  const path = errorPath(error);
  path && debug('path', path);
  debug('-------------------\n');
};

const createGraphQLErrorFormatter = (req?: Request) => (
  error: GraphQLError
) => {
  logGraphQLError(req, error);

  const isUserError = error.originalError
    ? error.originalError[IsUserError]
    : error[IsUserError];

  let sentryId = 'ID only generated in production';
  if (!isUserError) {
    if (process.env.NODE_ENV === 'production') {
      sentryId = Raven.captureException(
        error,
        req && Raven.parsers.parseRequest(req)
      );
    }
  }

  return {
    message: isUserError ? error.message : `Internal server error: ${sentryId}`,
    // Hide the stack trace in production mode
    stack:
      !process.env.NODE_ENV === 'production' ? error.stack.split('\n') : null,
  };
};

export default createGraphQLErrorFormatter;
