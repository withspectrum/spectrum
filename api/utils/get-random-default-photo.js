// @flow
const random = require('faker/lib/random');

// Helper function to get a random profile and cover photo
const PALETTE = [
  'blue',
  'blush',
  'cool',
  'green',
  'orange',
  'peach',
  'pink',
  'red',
  'teal',
  'violet',
];
export default () => {
  const color = random.arrayElement(PALETTE);
  return {
    profilePhoto: `/default_images/profile-${color}.png`,
    coverPhoto: `/default_images/cover-${color}.svg`,
  };
};
