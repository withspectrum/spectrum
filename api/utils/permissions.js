// @flow
import UserError from './UserError';
import type { GraphQLContext } from '../';
import type { DBChannel, DBCommunity, DBUser } from 'shared/types';
import {
  COMMUNITY_SLUG_BLACKLIST,
  CHANNEL_SLUG_BLACKLIST,
} from 'shared/slug-blacklists';
import { getThreadById } from '../models/thread';

export const isAdmin = (id: string): boolean => {
  const admins = [
    'gVk5mYwccUOEKiN5vtOouqroGKo1',
    '01p2A7kDCWUjGj6zQLlMQUOSQL42',
    'VToKcde16dREgDkXcDl3hhcrFN33',
  ];
  return admins.indexOf(id) > -1;
};

export const communitySlugIsBlacklisted = (slug: string): boolean => {
  return COMMUNITY_SLUG_BLACKLIST.indexOf(slug) > -1;
};

export const channelSlugIsBlacklisted = (slug: string): boolean => {
  return CHANNEL_SLUG_BLACKLIST.indexOf(slug) > -1;
};

// prettier-ignore
export const isAuthedResolver = (resolver: Function) => (obj: any, args: any, context: GraphQLContext, info: any) => {
  if (!context.user || !context.user.id || context.user.bannedAt) {
    return new UserError('You must be signed in to do this')
  }

  return resolver(obj, args, context, info)
}

// prettier-ignore
const channelExists = async (channelId: string, loaders: any): Promise<?DBChannel> => {
  const channel = await loaders.channel.load(channelId);
  if (!channel || channel.deletedAt) return null;
  return channel;
};

// prettier-ignore
const communityExists = async (communityId: string, loaders: any): Promise<?DBCommunity> => {
  const community = await loaders.community.load(communityId);
  if (!community || community.deletedAt) return null;
  return community;
};

// prettier-ignore
export const canAdministerChannel = async (userId: string, channelId: string, loaders: any) => {
  if (!userId || !channelId) return false;

  const channel = await channelExists(channelId, loaders);
  if (!channel) return false;

  const [communityPermissions, channelPermissions] = await Promise.all([
    loaders.userPermissionsInCommunity.load([userId, channel.communityId]),
    loaders.userPermissionsInChannel.load([userId, channelId]),
  ]);

  if (!communityPermissions) return false;
  if (communityPermissions.isOwner || communityPermissions.isModerator)
    return true;
  if (!channelPermissions) return false;
  if (channelPermissions.isOwner) return true;

  return false;
};

// prettier-ignore
export const canModerateChannel = async (userId: string, channelId: string, loaders: any) => {
  if (!userId || !channelId) return false;

  const channel = await channelExists(channelId, loaders);
  if (!channel) return false;

  const [communityPermissions, channelPermissions] = await Promise.all([
    loaders.userPermissionsInCommunity.load([userId, channel.communityId]),
    loaders.userPermissionsInChannel.load([userId, channelId]),
  ]);

  if (!communityPermissions) return false;
  if (communityPermissions.isOwner || communityPermissions.isModerator)
    return true;
  if (!channelPermissions) return false;
  if (channelPermissions.isOwner || channelPermissions.isModerator) return true;

  return false;
};

// prettier-ignore
export const canAdministerCommunity = async (userId: string, communityId: string, loaders: any) => {
  if (!userId || !communityId) return false;

  const community = await communityExists(communityId, loaders);
  if (!community) return false;

  const communityPermissions = await loaders.userPermissionsInCommunity.load([
    userId,
    communityId,
  ]);

  if (communityPermissions && communityPermissions.isOwner) return true;
  return false;
};

// prettier-igore
export const canModerateCommunity = async (
  userId: string,
  communityId: string,
  loaders: any
) => {
  if (!userId || !communityId) return false;

  const community = await communityExists(communityId, loaders);
  if (!community) return false;

  const communityPermissions = await loaders.userPermissionsInCommunity.load([
    userId,
    communityId,
  ]);

  if (!communityPermissions) return false;
  if (communityPermissions.isOwner || communityPermissions.isModerator)
    return true;
  return false;
};

// prettier-ignore
export const canViewCommunity = async (user: DBUser, communityId: string, loaders: any) => {
  if (!communityId) return false;

  const community = await communityExists(communityId, loaders);
  if (!community) return false;

  if (!community.isPrivate) return true

  if (!user) return false

  const communityPermissions = await loaders.userPermissionsInCommunity.load([
    user.id,
    communityId,
  ]);

  if (!communityPermissions) return false;
  if (!communityPermissions.isMember) return false
  
  return true;
}

export const canViewThread = async (
  userId: string,
  threadId: string,
  loaders: any
) => {
  const thread = await getThreadById(threadId);

  if (!thread || thread.deletedAt) return false;

  const [
    channel,
    community,
    channelPermissions,
    communityPermissions,
  ] = await Promise.all([
    loaders.channel.load(thread.channelId),
    loaders.community.load(thread.communityId),
    loaders.userPermissionsInChannel.load([userId, thread.channelId]),
    loaders.userPermissionsInCommunity.load([userId, thread.communityId]),
  ]);

  if (!channel.isPrivate && !community.isPrivate) return true;
  if (channel.isPrivate) return channelPermissions.isMember;
  if (community.isPrivate) return communityPermissions.isMember;
  return false;
};

export const canViewDMThread = async (
  userId: string,
  threadId: string,
  loaders: any
) => {
  if (!userId) return false;

  const thread = await loaders.directMessageParticipants.load(threadId);

  if (!thread || !thread.reduction || thread.reduction.length === 0)
    return false;

  const participants = thread.reduction;

  const ids = participants.map(({ userId }) => userId);

  if (ids.indexOf(userId) === -1) return false;

  return true;
};

// prettier-ignore
export const canViewChannel = async (user: DBUser, channelId: string, loaders: any) => {
  if (!channelId) return false;

  const channel = await channelExists(channelId, loaders);
  if (!channel) return false;

  const community = await communityExists(channel.communityId, loaders);
  if (!community) return false

  if (!channel.isPrivate && !community.isPrivate) return true

  if (!user) return false

  const [
    communityPermissions,
    channelPermissions
  ] = await Promise.all([
    loaders.userPermissionsInCommunity.load([
      user.id,
      community.id,
    ]),
    loaders.userPermissionsInChannel.load([
      user.id,
      channel.id,
    ])
  ])

  if (channel.isPrivate && !channelPermissions) return false
  if (community.isPrivate && !communityPermissions) return false
  if (channel.isPrivate && !channelPermissions.isMember) return false
  if (community.isPrivate && !communityPermissions.isMember) return false
  
  return true
}
