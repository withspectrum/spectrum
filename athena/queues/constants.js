// @flow
// A constant time buffer that determines the range in which we bundle notifications
export const TIME_BUFFER = 1800000; // 30 minutes
// A list of queue IDs so we don't accidentally mismatch the string constant

// user generated notifications
export const MESSAGE_NOTIFICATION = 'message notification';
export const MENTION_NOTIFICATION = 'mention notification';
export const DIRECT_MESSAGE_NOTIFICATION = 'direct message notification';
export const REACTION_NOTIFICATION = 'reaction notification';
export const CHANNEL_NOTIFICATION = 'channel notification';
export const COMMUNITY_NOTIFICATION = 'community notification';
export const THREAD_NOTIFICATION = 'thread notification';
export const COMMUNITY_INVITE_NOTIFICATION = 'community invite notification';
export const SEND_COMMUNITY_INVITE_EMAIL = 'send community invite email';
export const SLACK_IMPORT = 'slack import';
export const SEND_NEW_MESSAGE_EMAIL = 'send new message email';
export const SEND_MENTION_THREAD_NOTIFICATION_EMAIL =
  'send thread mention email';
export const SEND_MENTION_MESSAGE_NOTIFICATION_EMAIL =
  'send message mention email';
export const SEND_NEW_DIRECT_MESSAGE_EMAIL = 'send new direct message email';
export const SEND_THREAD_CREATED_NOTIFICATION_EMAIL =
  'send thread created notification email';

// invoices and recurring payments
export const PRO_INVOICE_PAID_NOTIFICATION = 'pro invoice paid notification';
export const COMMUNITY_INVOICE_PAID_NOTIFICATION =
  'community invoice paid notification';
export const SEND_PRO_INVOICE_RECEIPT_EMAIL = 'send pro invoice receipt email';
export const SEND_COMMUNITY_INVOICE_RECEIPT_EMAIL =
  'send community invoice receipt email';
