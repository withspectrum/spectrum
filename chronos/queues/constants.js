// @flow

// counts for processing
// the thread must have at least # total messages
export const MIN_TOTAL_MESSAGE_COUNT = 5;
// # of the total messages must have been sent in the past week
export const MIN_NEW_MESSAGE_COUNT = 5;
// # only show the top # threads per channel
export const MAX_THREAD_COUNT_PER_CHANNEL = 10;
// don't send the digest if the email will have less than # total threads to show
export const MIN_THREADS_REQUIRED_FOR_DIGEST = 5;
// cap the digest at # threads
export const MAX_THREAD_COUNT_PER_DIGEST = 10;
// upsell communities to join if the user has joined less than # communities
export const COMMUNITY_UPSELL_THRESHOLD = 5;

// generate a score for each thread based on the total number of messages and number of new messages
// new messages rank higher in order to devalue old threads that have a large amount of old messages (like pinned posts)
// the end weekly digest will have threads sorted by the weight of (TOTAL * WEIGHT) + (NEW * WEIGHT)
export const TOTAL_MESSAGE_COUNT_WEIGHT = 0.1;
export const NEW_MESSAGE_COUNT_WEIGHT = 1.5;

/*
  Example weighting:
  A thread with 150 messages, where 5 are new this week: 22.5
  A thread with 10 total messages, where all 10 are new this week: 16
  A thread with 25 total messages, where 10 are old and 15 are new this week: 25
*/

// queues
export const SEND_DIGEST_EMAIL = 'send digest email';
export const PROCESS_WEEKLY_DIGEST_EMAIL = 'process weekly digest email';
export const PROCESS_DAILY_DIGEST_EMAIL = 'process daily digest email';
