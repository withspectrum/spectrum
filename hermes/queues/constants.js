// @flow
const IS_PROD = !process.env.FORCE_DEV && process.env.NODE_ENV === 'production';

export const SEND_NEW_MESSAGE_EMAIL = 'send new message email';
export const SEND_NEW_DIRECT_MESSAGE_EMAIL = 'send new direct message email';
export const SEND_COMMUNITY_INVITE_EMAIL = 'send community invite email';
export const SEND_NEW_USER_WELCOME_EMAIL = 'send new user welcome email';
export const SEND_NEW_COMMUNITY_WELCOME_EMAIL =
  'send new community welcome email';

export const SEND_COMMUNITY_INVOICE_RECEIPT_EMAIL =
  'send community invoice receipt email';
export const SEND_COMMUNITY_PAYMENT_SUCCEEDED_EMAIL =
  'send community payment succeeded email';
export const SEND_COMMUNITY_PAYMENT_FAILED_EMAIL =
  'send community payment failed email';
export const SEND_COMMUNITY_CARD_EXPIRING_WARNING_EMAIL =
  'send community card expiring warning email';

export const SEND_PRO_INVOICE_RECEIPT_EMAIL = 'send pro invoice receipt email';
export const SEND_THREAD_CREATED_NOTIFICATION_EMAIL =
  'send thread created notification email';
export const SEND_DIGEST_EMAIL = 'send digest email';
export const SEND_EMAIL_VALIDATION_EMAIL = 'send email validation email';
export const SEND_ADMINISTRATOR_EMAIL_VALIDATION_EMAIL =
  'send administrator email validation email';
export const SEND_ADMIN_COMMUNITY_CREATED_EMAIL = 'admin community created';
export const SEND_NEW_MENTION_THREAD_EMAIL = 'send thread mention email';
export const SEND_NEW_MENTION_MESSAGE_EMAIL = 'send message mention email';
export const SEND_ADMIN_TOXIC_MESSAGE_EMAIL = 'admin toxic content email';
export const SEND_ADMIN_SLACK_IMPORT_PROCESSED_EMAIL =
  'admin slack import processed email';
export const SEND_ACTIVE_COMMUNITY_ADMIN_REPORT_EMAIL =
  'send active community admin report email';
export const SEND_PRIVATE_CHANNEL_REQUEST_SENT_EMAIL =
  'send request join private channel email';
export const SEND_PRIVATE_CHANNEL_REQUEST_APPROVED_EMAIL =
  'send private channel request approved email';
export const SEND_ADMIN_USER_SPAMMING_THREADS_NOTIFICATION_EMAIL =
  'send admin user spamming threads notification email';

export const NEW_MESSAGE_TEMPLATE = IS_PROD ? 2266041 : 3788381;
export const NEW_MENTION_THREAD_TEMPLATE = IS_PROD ? 3776541 : 3844623;
export const NEW_MENTION_MESSAGE_TEMPLATE = IS_PROD ? 3844364 : 3844624;
export const NEW_DIRECT_MESSAGE_TEMPLATE = 2911541;
export const NEW_USER_WELCOME_TEMPLATE = 2462726;
export const COMMUNITY_INVITE_TEMPLATE = 2302401;
export const NEW_COMMUNITY_WELCOME_TEMPLATE = 2600301;

export const COMMUNITY_INVOICE_RECEIPT_TEMPLATE = IS_PROD ? 2647483 : 12345;
export const COMMUNITY_PAYMENT_SUCCEEDED_TEMPLATE = IS_PROD ? 5436583 : 5360481;
export const COMMUNITY_PAYMENT_FAILED_TEMPLATE = IS_PROD ? 5436762 : 5359821;
export const COMMUNITY_CARD_EXPIRING_WARNING_TEMPLATE = IS_PROD
  ? 5436582
  : 5381761;

export const PRO_INVOICE_RECEIPT_TEMPLATE = 3037461;
export const NEW_THREAD_CREATED_TEMPLATE = IS_PROD ? 2713302 : 3786781;
export const DIGEST_TEMPLATE = IS_PROD ? 3071361 : 4165801;
export const DEBUG_TEMPLATE = 3374841;
export const EMAIL_VALIDATION_TEMPLATE = 3578681;
export const ADMINISTRATOR_EMAIL_VALIDATION_TEMPLATE = IS_PROD
  ? 5796281
  : 4952721;

export const ADMIN_COMMUNITY_CREATED_TEMPLATE = 3037441;
export const ADMIN_TOXIC_MESSAGE_TEMPLATE = 3867921;
export const ADMIN_SLACK_IMPORT_PROCESSED_TEMPLATE = 3934361;
export const ADMIN_ACTIVE_COMMUNITY_REPORT_TEMPLATE = 3947362;
export const ADMIN_USER_SPAMMING_THREADS_NOTIFICATION_TEMPLATE = 5736761;

export const PRIVATE_CHANNEL_REQUEST_SENT_TEMPLATE = IS_PROD
  ? 4550702
  : 4543221;
export const PRIVATE_CHANNEL_REQUEST_APPROVED_TEMPLATE = IS_PROD
  ? 4550804
  : 4543861;

// types used to generate unsubscribe tokens
export const TYPE_DAILY_DIGEST = 'dailyDigest';
export const TYPE_WEEKLY_DIGEST = 'weeklyDigest';
export const TYPE_NEW_THREAD_CREATED = 'newThreadCreated';
export const TYPE_NEW_MESSAGE_IN_THREAD = 'newMessageInThreads';
export const TYPE_NEW_DIRECT_MESSAGE = 'newDirectMessage';
export const TYPE_NEW_MENTION = 'newMention';
export const TYPE_MUTE_CHANNEL = 'muteChannel';
export const TYPE_MUTE_COMMUNITY = 'muteCommunity';
export const TYPE_MUTE_THREAD = 'muteThread';
export const TYPE_MUTE_DIRECT_MESSAGE_THREAD = 'muteDirectMessageThread';
