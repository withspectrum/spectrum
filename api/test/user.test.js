// @flow
import { request } from './utils';
import { MAX_ID } from '../migrations/seed/default/constants';

describe('queries', () => {
  it('should fetch a user', async () => {
    const query = /* GraphQL */ `
      {
        user(id: "${MAX_ID}") {
          id
          name
          description
          website
          username
          email
          providerId
          createdAt
        }
      }
    `;

    expect.assertions(1);
    const result = await request(query);

    expect(result).toMatchSnapshot();
  });

  it('should return null for a non-existant id', async () => {
    const query = /* GraphQL */ `
      {
        user(id: "non-existant") {
          id
          username
        }
      }
    `;

    expect.assertions(1);
    const result = await request(query);

    expect(result).toMatchSnapshot();
  });

  it('should fetch a users communities', async () => {
    const query = /* GraphQL */ `
      {
        user(id: "${MAX_ID}") {
          communityConnection {
            edges {
              node {
                name
              }
            }
          }
        }
      }
    `;

    expect.assertions(1);
    const result = await request(query);

    expect(result).toMatchSnapshot();
  });

  describe.skip('everything', () => {
    it('should return the latest thread', async () => {
      const query = /* GraphQL */ `
        {
          user(id: "${MAX_ID}") {
            everything(first: 1) {
              pageInfo {
                hasNextPage
              }
              edges {
                node {
                  id
                }
              }
            }
          }
        }
      `;
      expect.assertions(1);
      const result = await graphql(schema, query);

      expect(result).toMatchSnapshot();
    });

    it('should paginate based on the after property', async () => {
      const query = /* GraphQL */ `
        {
          user(id: "${MAX_ID}") {
            everything(first: 1, after: "c2Vjb25kLXN0b3J5") {
              pageInfo {
                hasNextPage
              }
              edges {
                node {
                  id
                }
              }
            }
          }
        }
      `;
      expect.assertions(1);
      const result = await graphql(schema, query);

      expect(result).toMatchSnapshot();
    });

    it('should handle first being set to 0 correctly', async () => {
      const noCursorQuery = /* GraphQL */ `
        {
          user(id: "${MAX_ID}") {
            everything(first: 0) {
              pageInfo {
                hasNextPage
              }
              edges {
                node {
                  id
                }
              }
            }
          }
        }
      `;
      const cursorQuery = /* GraphQL */ `
        {
          user(id: "${MAX_ID}") {
            everything(first: 0, after: "Zmlyc3Qtc3Rvcnk=") {
              pageInfo {
                hasNextPage
              }
              edges {
                node {
                  id
                }
              }
            }
          }
        }
      `;
      expect.assertions(2);
      const result = await graphql(schema, noCursorQuery);

      expect(result).toMatchSnapshot();

      const nextResult = await graphql(schema, cursorQuery);

      expect(nextResult).toMatchSnapshot();
    });
  });
});
