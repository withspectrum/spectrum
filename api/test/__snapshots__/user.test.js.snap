// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queries should fetch a user 1`] = `
Object {
  "data": Object {
    "user": Object {
      "createdAt": "2016-12-31T23:00:00.000Z",
      "description": "Makes styled-components, react-boilerplate and micro-analytics ðŸ’… Speciality coffee geek, skier, traveller â˜•",
      "email": null,
      "id": "1",
      "name": "Max Stoiber",
      "providerId": "2451223458",
      "username": "mxstbr",
      "website": "https://mxstbr.com",
    },
  },
}
`;

exports[`queries should fetch a users communities 1`] = `
Object {
  "data": Object {
    "user": Object {
      "communityConnection": Object {
        "edges": Array [
          Object {
            "node": Object {
              "name": "Payments",
            },
          },
          Object {
            "node": Object {
              "name": "Spectrum",
            },
          },
        ],
      },
    },
  },
}
`;

exports[`queries should return null for a non-existant id 1`] = `
Object {
  "data": Object {
    "user": null,
  },
}
`;
